### DISCUSSION TITLE
Is there a way of upserting fixed relations? reflect on existing relations?

### DISCUSSION THREAD

**Discussion by Madd0g** [cozodb/cozo#50] [Q&A]
Sorry if some of these are beginner question, I just started learning.

1. I tried to `:create` the same relation more than once and it failed, I'm wondering if there's a common pattern (I suppose that works with transactions) to upsert a fixed relation?

Would appreciate an example of how to also input data at the same time as upserting.

2. I saw that there's `::relations` and `::columns <REL>` - but if the database grows I want a way to get existing relations selectively, are there other ways of reflecting on what's in the database?

3. Also, are there any other tutorials/materials other than the docs page? or examples of projects using cozo? I'm more interested in the graph-db aspects, the airports example doesn't show a lot of arbitrary connections you might see in a graph-db. I'm new to datalog too so having just one complex example is limiting my ability to learn :)

Thanks!

**Comment by zh217** [comment]
For 1., there is `:replace` in addition to `:create`, but it is not an upsert, it is replacing the old relation completely if it exists.

There is a way to do true upsert in a single transaction: you can use either query chaining https://docs.cozodb.org/en/latest/stored.html#chaining-queries (specifically, `%ignore_error <query expression>`), or multistatement transaction https://docs.cozodb.org/en/latest/stored.html#multi-statement-transaction .

For 2., no currently there is no other way. But getting metadata should be extremely fast. Is there any reason that filtering within your host language is infeasible?

For 3., unfortunately no, since Cozo is pretty new. But there are more material on the web if you search for Datalog (still not a lot, and you will need to translate the syntax).

**Reply by Madd0g** [reply]
1. Thanks, I think I get it.

2. I guess I'm wondering about introspection in general, given that I know the named relations and their columns, can I do basic introspection of the data? like `select * FROM relation` and then keep drilling into what is there.

3. I tried to do just that. I have little experience with prolog, so some concepts in datalog were easier to grasp than others, but cozo is another conceptual jump and some things are hard to wrap my head around.

Thanks you!
